app/code/community/Bolt/Boltpay/controllers/ShippingController.php

```diff
@@ -168,39 +171,69 @@ public function indexAction()
-        $quote->getBillingAddress()->addData($address_data);
+        $shippingAddress = $quote->getShippingAddress();
+        if (empty($shippingAddress->getCountryId()))
+        {
+            $cacheKey = $this->generateCacheKey($quote->getId(), $shippingAddress, $quote->getGrandTotal());
+            ////////////////////////////////////////////////////////////////////////
+            // Clear previously set estimates.  This helps if this
+            // process fails or is aborted, which will force the actual index action
+            // to get fresh data instead of reading from the session cache
+            ////////////////////////////////////////////////////////////////////////
+            Mage::app()->getCache()->remove("quote_location_" . $cacheKey);
+            Mage::app()->getCache()->remove("quote_shipping_and_tax_estimate_" . $cacheKey);
+            ////////////////////////////////////////////////////////////////////////
 
-        try {
-            $estimate_response = Mage::helper('boltpay/api')->getShippingAndTaxEstimate($quote);
-        } catch (Exception $e) {
-            $estimate_response = null;
+            $request_json = file_get_contents('php://input');
+            $request_data = json_decode($request_json);
+
+            $address_data = array(
+                'city'          => $request_data->city,
+                'region'        => $request_data->region_code,
+                'region_name'   => $request_data->region_name,
+                'postcode'      => $request_data->zip_code,
+                'country_id'    => $request_data->country_code
+            );
+            $quote->getShippingAddress()->addData($address_data);
+            $quote->getBillingAddress()->addData($address_data);
+
+            try {
+                $estimate_response = Mage::helper('boltpay/api')->getShippingAndTaxEstimate($quote);
+            } catch (Exception $e) {
+                $estimate_response = null;
+            }
+
+            Mage::app()->getCache()->save(serialize($address_data), "quote_location_" . $cacheKey);
+            Mage::app()->getCache()->save(serialize($estimate_response), "quote_shipping_and_tax_estimate_" . $cacheKey);
+        } else {
+            $address_data = array();
         }
 
-        Mage::app()->getCache()->save(serialize($address_data), "quote_location_".$quote->getId());
-        Mage::app()->getCache()->save(serialize($estimate_response), "quote_shipping_and_tax_estimate_".$quote->getId());
+        $response = Mage::helper('core')->jsonEncode(array('address_data' => $address_data));
+        $this->getResponse()->setHeader('Content-type', 'application/json');
+        $this->getResponse()->setBody($response);
+    }
+
+    /**
+     * @param $quoteId
+     * @param $_shippingAddress
+     * @param $totalAmount
+     * @return string
+     */
+    public function generateCacheKey($quoteId, $_shippingAddress, $totalAmount)
```

Comment:
sorry I'm keep saying this but I prefer we include skus here. Think about $50 gift card vs $50 chair. Price is same, shipping cost would be different.